chainerrl/agents/dqfd.py

```diff
@@ -0,0 +1,679 @@
+        self.average_loss_1step += (1 - self.average_loss_decay) * \
+            float(loss_q_1step.array)
+
+        self.average_loss_nstep *= self.average_loss_decay
+        self.average_loss_nstep += (1 - self.average_loss_decay) * \
+            float(loss_q_nstep.array)
+
+        self.average_loss_supervised *= self.average_loss_decay
+        self.average_loss_supervised += (1 - self.average_loss_decay) * \
+            float(loss_supervised.array)
+
+        if len(err_demo):
+            self.average_demo_td_error *= self.average_loss_decay
+            self.average_demo_td_error += (1 - self.average_loss_decay) * \
+                np.mean(err_demo)
+
+        if len(err_agent):
+            self.average_agent_td_error *= self.average_loss_decay
+            self.average_agent_td_error += (1 - self.average_loss_decay) * \
+                np.mean(err_agent)
+
+    def _compute_y_and_ts(self, exp_batch):
+        """Compute output and targets
+
+        Changes from DQN:
+            Cache qout for the supervised loss later
+            Calculate both 1-step and n-step targets
+        """
+        batch_size = exp_batch['reward'].shape[0]
+
+        # Compute Q-values for current states
+        batch_state = exp_batch['state']
+        qout = self.model(batch_state)
+
+        # Caches Q(s) for use in supervised demo loss
+        self.qout = qout
+
+        batch_actions = exp_batch['action']
+        batch_q = F.reshape(qout.evaluate_actions(
+            batch_actions), (batch_size, 1))
+
+        with chainer.no_backprop_mode():
+            # Calculate Double DQN targets
+            batch_q_target = F.reshape(
+                self._compute_target_values(exp_batch),
+                (batch_size, 1))
+        return batch_q, batch_q_target
+
+    def _compute_ddqn_losses(self, exp_batch, errors_out=None):
+        """Compute the Q-learning losses for a batch of experiences
+
+        Args:
+          exp_batch (dict): A dict of batched arrays of transitions
+        Returns:
+          Computed loss from the minibatch of experiences
+        """
+        y, t = self._compute_y_and_ts(exp_batch)
+
+        del errors_out[:]
+        delta = F.absolute(y - t)
+        if delta.ndim == 2:
+            delta = F.sum(delta, axis=1)
+        delta = cuda.to_cpu(delta.array)
+        for e in delta:
+            errors_out.append(e)
+
+        is_1_step = self.xp.abs(1. - exp_batch["is_n_step"])
+        loss_1step = compute_weighted_value_loss(
+            y, t, exp_batch['weights'],
+            mask=is_1_step,
+            clip_delta=self.clip_delta,
+            batch_accumulator=self.batch_accumulator)
+        loss_nstep = compute_weighted_value_loss(
+            y, t, exp_batch['weights'],
+            mask=exp_batch["is_n_step"],
+            clip_delta=self.clip_delta,
+            batch_accumulator=self.batch_accumulator)
+        return loss_nstep, loss_1step
+
+    def act_and_train(self, obs, reward):
```

Comment:
Seems like inheriting this from DQN would do since there doesn't seem to be any change

Reply:
Absolutely right. Thanks!